### ISSUE TITLE
`Quantity.__setitem__` skips unit check when given a mask array

### ISSUE THREAD

**Issue by SimonHeybrock** [hgrecco/pint#1584]
This works:
```python
import numpy as np
import pint

masked = np.ma.array([1,2,], mask=[True, False])
np.ma.is_masked(masked)

q = pint.Quantity(np.ones(shape=(2, )), 'm')
q[:] = masked  # no unit, but accepted

masked_q = pint.Quantity(masked, 'mm')
q[:] = masked_q  # wrong unit, but not checked
```
Offending code is here: https://github.com/hgrecco/pint/blob/208d36a5bb051465296dad846296bb19628e2307/pint/facets/numpy/quantity.py#L246-L249

Not sure how this `np.ma.is_masked` was intended?

**Comment by hgrecco** [comment]
You can find the rationale here:

https://github.com/hgrecco/pint/commit/185ec6b81c1c6f57263733571face3ef672c4afd

If there is a better solution, we would accept a patch

**Comment by dopplershift** [comment]
The idea in that change is that masked should behave the same as nan. I have personally never had a problem with nan/masked essentially not having units.

**Comment by SimonHeybrock** [comment]
> The idea in that change is that masked should behave the same as nan. I have personally never had a problem with nan/masked essentially not having units.

But it is not masked, look at my example, there are two problems:

1. It checks if the new `value` is masked, but the magnitude is not masked, i.e., it "drops" the mask and uses that value.
2. `np.ma.is_masked` returns `True` if *any* of the elements is masked, but there can be many others that are not masked.

Shouldn't this rather check if the magnitude is masked at `key`?

Though even then I don't feel it is conceptually correct to ignore the units --- masked or not, wrong units indicates a bug in the user's code.

**Comment by dopplershift** [comment]
So the original intent was to allow users to mask a value without worrying about units (or, more importantly, allowing code that doesn't know about Pint to work without error or units). So you want to be able to do:

```python
import numpy as np
import pint

a = np.ma.array([1, 2, 3], mask=False)
q = pint.Quantity(a, 'm')
q[2] = np.ma.masked
```

and the same for NaN. In my mind, the primary purpose of pint is to make sure units are tracked and converted properly, and incidentally it errors out when it can't sensibly do that. Whatever the differences in units are, the result of converting NaN is still NaN. Same thing for masked. Yes, that can allow user bugs to pass, but it also *really* helps interoperability--and it won't result in any incorrect answers that you wouldn't have already gotten.

Now, with all that said, there is a bug there. When I made that patch, that branch is essentially explicitly there for the scalar case, and it's supposed to fall through the `TypeError` to handle the array case. I missed the fact that `np.ma.is_masked()` returns `True` if *any* element is masked. Whereas, before my patch, `math.isnan` only works on scalars. I'll put together a PR to fix that.

**Comment by dopplershift** [comment]
I think I have a reasonably clean fix in #1608.
